### ISSUE TITLE
NuShell Support

### ISSUE THREAD

**Issue by f1rstC0de** [warpdotdev/Warp#2038]
### Describe the solution you'd like?

NuShell Support (autocomplit, etc)

https://github.com/nushell/nushell
https://www.nushell.sh/

### How important is this feature to you?

5

**Comment by dannyneira** [comment]
Thanks for this feature request! Please +1 this request as it helps us gauge interest. We will post any updates here.

**Comment by ammkrn** [comment]
> Thanks for this feature request! Please +1 this request as it helps us gauge interest. We will post any updates here.

@dannyneira Can you give the people who thumbsup'd OP a little bit of information on what's needed to make this work, or is it all on Warp's end?

**Comment by dannyneira** [comment]
Yes we'd have to correctly pass nushell stdin, stdout and stderr to Warp's input editor to add support. Currently only bash, zsh and fish are supported, with zsh being the default.

**Comment by dannyneira** [comment]
Thanks for this feature request!

To anyone else interested in this feature, please add a 👍 to the original post at the top to signal that you want this feature, and subscribe if you'd like to be notified.

Also, please avoid comments like "+1" or "👍" on the thread to reduce spamming folks' inboxes.

**Comment by daUnknownCoder** [comment]
Damn, I saw the `closed #880 as completed`, I thought all shell support were added in 1 go 😂

**Comment by louiss0** [comment]
Can you guys please support Nushell? I has it's own auto completions! All you need to do is Warpify it!

**Comment by louiss0** [comment]
> > Hey,
> > 
> > I don't mean to make noise here and I don't know completely what your open-source policy is but I just wanted to let you know that I would be willing to work on this if you are indeed looking for member of community to take this up.
> 
> pls do ^^

I wish I could help, but I do know the language enough to know that you need a plugin!

**Comment by theslyprofessor** [comment]
Hello! I'm happy to help as well. I believe this will be a game changer. Nushell excels at efficiently piping outputs, making it ideal for powering an agent terminal like Warp.

**Comment by archcorsair** [comment]
> [@theslyprofessor](https://github.com/theslyprofessor) could you give an example on how NuShell helps with agentic terminals specifically?

If you don't mind I'd like to give my personal opinion on this:
Nushell can make agentic terminals more powerful by enabling advanced data processing, structured output and automation.

(I used an LLM to structure my brain dump)

# 1. Human-Readable, Structured Output ￼
NuShell outputs data in tables, lists, and other structured formats (like JSON, YAML, and TOML) by default. This makes it much easier for both humans and agents to parse, transform, and act on command results compared to traditional plain-text output.

**Example:**
Listing files with metadata
```nushell
ls | where size > 1mb | sort-by size
```
This produces a clean, filterable table, ideal for further processing or display in a UI.

# 2. Powerful Data Processing Primitives ￼
NuShell provides built-in commands (primitives) for filtering, sorting, grouping, and transforming data—no need for external tools like grep, awk, or jq.

**Example:**
Extracting and summarizing log data
```nushell
open logs.json | where level == "error" | group-by source | count
```
This can quickly give an agent a summary of error sources, ready for reporting or alerting.

# 3. Seamless Data Format Conversion ￼
NuShell can easily convert between formats (CSV, JSON, YAML, etc.), which is **invaluable** for agentic workflows that need to interface with APIs, config files, or other tools.
**Example:**
Converting CSV to JSON for an API call:
```nushell
open data.csv | to json | save data.json
```

# 4. Composable Pipelines for Automation ￼
NuShell’s pipeline model allows chaining commands in a way that’s both readable and robust, making it easier for agents to automate complex tasks.

**Example:**
Automating a backup and reporting process:
```nushell
ls ~/Documents | where modified > (date now - 7day) | zip backup.zip | save log.txt
```

# 5. Enhanced Scripting for Agents ￼
NuShell scripts are more expressive and safer than traditional shell scripts, reducing the risk of errors and making it easier for agents to generate or modify scripts on the fly.

In summary, NuShell’s structured data handling, rich primitives, and format flexibility make it a natural fit for agentic terminals that need to interpret, transform, and act on data efficiently.

Thanks for reading if you got this far ❤ Nushell is an important part of my daily workflow and would love to have it integrated into Warp at some point in the future

**Comment by vivekanandan-ks** [comment]
If someone really sees the power of nushell they'll understand how much great it'll be with warp or any terminal. It's just really good. Sad to see it not supported yet even after 2.5+ years.

**Comment by brendon-felix** [comment]
I won't start using Warp until Nushell is supported, however I do think there are some important considerations:

1. Nushell is still under active development with breaking changes between versions. Commands, flags, and syntax structures are still being added, removed and changed constantly.
2. Existing Nushell-related training data for LLMs is sparse and outdated, meaning code generation is currently unreliable.
3. The number of Nushell users, while significant and increasing, is still quite low compared to currently supported shells. This means adding Nushell support may not increase Warp's userbase as much as adding support for other shells first.

Despite these considerations, I do want to say I *love* Nushell for it's look and feel, structured data approach, language design inspired by rust, and it's cross-platform support. I hope to see it get first-class support in Warp.

**Comment by archcorsair** [comment]
I believe this recent nushell addition could be  relevant to this issue: https://github.com/nushell/nushell/pull/16693

**Comment by Miuler** [comment]
Please add support for nushell

**Comment by archcorsair** [comment]
> I'm really puzzled by the fact the dev team doesn't seem to realize the incredible synergy between Nushell and AI. Nushell comes with its own MCP and has a LSP. If anything Nushell should be marketed as their primary shell considering the direction they have taken over the last year with AI.
> 
> (Dev team there is really something big you are missing out on here...)
> 
> References:
> 
> * [nushell.sh/blog/2025-10-15-nushell_v0_108_0.html#nushell-mcp-server-16693-16758-16839-16857-toc](https://www.nushell.sh/blog/2025-10-15-nushell_v0_108_0.html#nushell-mcp-server-16693-16758-16839-16857-toc)

I’d like to give a developer’s perspective here. I’m also not using Warp largely because of the lack of Nushell support, but at the same time I _totally_ understand why it’s taking so long or may simply not be a priority.

Nushell isn’t just “another shell” in the way bash vs zsh are. It intentionally steps away from POSIX semantics and replaces a lot of fundamental assumptions: structured data instead of text streams, different built-ins, different piping model, different completion and error semantics. That is a huge architectural divergence.

From the outside, Warp likely has a fairly deep shell integration layer that assumes POSIX-style behavior. Supporting bash, zsh, and fish is hard but tractable because they still share a lot of underlying concepts. Nushell is a completely different execution and interaction model, so adding support isn’t a toggle. It is closer to building a parallel integration path.

I’d love to see it happen. I just don’t think it’s fair to frame this as the dev team “missing something obvious.” This is one of those cases where the technical reality is genuinely hard.

**Comment by ramarivera** [comment]
2026 and I still can't come back to warp no matter how much I want to :(

**Comment by tukanoidd** [comment]
> That's why good abstractions matter. Besides, they can't assume POSIX compliance since they support fish, which isn't.

\+ like, the engine is fully embeddable as a library, use it in my personal project, so it technically should've been the easiest to integrate into a Rust codebase too, no std::process::Command shenannigans and dealing with process input/output piping

**Comment by POLARIS0824** [comment]
Still no support for nushell?😭

**Comment by freepicheep** [comment]
Hmm, now that Warp is open source we can point an agent at this issue.🤔👀

**Comment by jexroid** [comment]
the thing is, will they accept the pull req? @freepicheep

**Comment by freepicheep** [comment]
> the thing is, will they accept the pull req? @freepicheep 

Who knows. Hopefully. However, I guess that's what personal forks are for.🤷‍♂️
